### ISSUE TITLE
Builder Metadata Format

### ISSUE THREAD

**Issue by sclevine** [buildpack/pack#251]
**Given** I have a builder.toml with buildpacks `buildpack.id@1.2.4` and `other.id@3.4.5`
**When** I create a builder
**Then** the builder contains the following labels:

-  `io.buildpacks.buildpack.layers` (`version`s are not optional):
```json
{
  "buildpack.id": {
    "1.2.4": {
      "layerDigest": "sha256:8abc...",
      "order": [
        {"group": [{"id": "other.id", "version": "3.4.5"}]}
      ]
    }
  },
  "other.id": {
    "3.4.5": {
      "layerDigest": "sha256:9acc..."
    }
  }
}
```
- `io.buildpacks.buildpack.order` (builder-only, `version`s are optional):
```json
[
   {"group": [{"id": "buildpack.id", "version": "1.2.3"}]},
   {"group": [{"id": "other.id"}]}
]
```

- `io.buildpacks.builder.metadata` with current metadata contents (for backwards compatibility)

**Notes**
- Moved buildpackage requirements to #289

**Comment by ekcasey** [comment]
@sclevine right now we have a structured `io.buildpacks.builder.metadata` label, with both `buildpacks` and `order` in it. I don't have any issues with format of the specific pieces of data, but I don't see the motivation for breaking the data out into these these two new labels.

I like that `io.buildpacks.builder.metadata` indicates that the image is in fact a builder, `io.buildpacks.layers` seems ambiguous enough that it could refer to many things in the ecosystem

**Comment by ameyer-pivotal** [comment]
@ekcasey @sclevine So (for builders) does this story really just boil down to adding the `order` key to a given buildpack in the metadata?

**Comment by sclevine** [comment]
@ekcasey @ameyer-pivotal edited to clarify the versions in the order metadata should be optional.

**Comment by ameyer-pivotal** [comment]
@sclevine I'd like to make a case for having order versions required:

- We will have all the information needed to insert it here (whether it comes from `builder.toml` or `--buildpack` flag).
- This label seems to serve somewhat as a record of the exact pieces built into the builder. We should be clear and exact everywhere possible.
- Optionality create ambiguity and code complexity

Could you elaborate on the need for making it optional?

**Comment by sclevine** [comment]
This is not a stylistic preference. Without optional versions in `io.buildpacks.buildpack.order`, it becomes impossible to add new buildpacks to the builder without a copy of the builder author's original order definition.

The optionality of the versions encodes information that is lost when the versions are resolved: the intended build configuration of the builder author. If you want to create another label that has them resolved, we could, but all the information is already encoded in the two labels described.

To address some of your other points:

> This label seems to serve somewhat as a record of the exact pieces built into the builder.

The `layers` label serves this purpose. The builder can contain other buildpacks that are not the mentioned in the order `label`.

> Optionality create ambiguity and code complexity

The functionality is already necessary to support optional versions in builder.toml.

**Comment by jromero** [comment]
Updated issue to reflect decision. **YES** on algorithm prepending.

**Comment by jromero** [comment]
@ekcasey / @sclevine what did we decide on the layers metadata? Are we using compressed, uncompressed or both when adding hashed to the label?

**Comment by ameyer-pivotal** [comment]
@jromero Compressed in both cases

**Comment by jromero** [comment]
@sclevine we've found that the layer digest is unreliable and completely dependent on the tool used for pushing the layers to the registry.

We discovered that the layer digests are different when pushing from `pack build ... --publish` vs `docker push`. This lead us down a rabbit hole which proved just how unreliable layer digests are and that we probably shouldn't be using them for tracking (metadata).

To further prove the point, here's a quick sample of the implications:

1. `pack create-builder index.docker.io/jromeropivot/builder-0325 --builder-config ~/dev/samples/builders/alpine/builder.toml  --publish --verbose`
2. `docker manifest inspect index.docker.io/jromeropivot/builder-0325`
3. (save output)
4. `docker pull index.docker.io/jromeropivot/builder-0325`
5. `docker push index.docker.io/jromeropivot/builder-0325` (note: no changes)
6. `docker manifest inspect index.docker.io/jromeropivot/builder-0325`
7. (save output)

**Result**
Layers added by pack change digests when comparing manifests

We could align `pack` to generate the same layer digests but we're not confident that every tool [(including GGCR)](https://github.com/google/go-containerregistry/issues/572) can guarantee this will be the case.
